### PR TITLE
Add selfhosted wakatime integration (wakapi)

### PR DESCRIPTION
**Pull request by MeerBiene**
This PR adds the option to use a selfhosted wakatime instance for the wakatime plugin. 

I tested it against this project called [Wakapi](https://github.com/muety/wakapi) which mimics the same endpoints as wakatime does, so its just a matter of using a different api url

### CODE REVIEW

File: source/plugins/wakatime/metadata.yml [hunk 1]
```diff
@@ -53,3 +52,22 @@ inputs:
...
+  # would be https://wakapi.dev
+  plugin_wakatime_apiurl:
+    description: Address where to reach your Wakapi instance
+    type: string
+    default: ""
+
+  plugin_wakatime_user:
+    description: Your Wakatime user on the selfhosted Wakapi instance
+    type: string
+    default: ""
```
Comment (lowlighter) [inline source/plugins/wakatime/metadata.yml:73]:
```suggestion
    default: .user.login
```
`.user.login` will automatically be parsed as current user

File: source/plugins/wakatime/index.mjs [hunk 2]
```diff
@@ -1,44 +1,56 @@
...
-            limit = void(limit)
-          const range = {"7":"last_7_days", "30":"last_30_days", "180":"last_6_months", "365":"last_year"}[days] ?? "last_7_days"
+    //Load inputs
+    let {
+      sections,
+      days,
+      limit,
+      apiurl,
+      wakapi = false,
+      user = "current",
```
Comment (lowlighter) [inline source/plugins/wakatime/index.mjs:15]: No need to set defaults on these, they're automatically set by `imports.metadata.plugins.wakatime.inputs` depending on what is defined in `metadata.yml` 👍

File: source/plugins/wakatime/index.mjs [hunk 3]
```diff
@@ -1,44 +1,56 @@
...
-            },
-            projects:stats.projects.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            languages:stats.languages.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            os:stats.operating_systems.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            editors:stats.editors.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-          }
+    //Querying api and format result (https://wakatime.com/developers#stats)
+    console.debug(`metrics/compute/${login}/plugins > wakatime > querying api`);
+    const {data: { data: stats }} = await imports.axios.get(
+      `${wakapi !== false ? apiurl : "http://wakatime.com"}/api/v1/users/${user}/stats/${range}?api_key=${token}`
```
Comment (lowlighter) [inline source/plugins/wakatime/index.mjs:29]:
```suggestion
      `${apiurl}/api/v1/users/${user}/stats/${range}?api_key=${token}`
```
Let's just default `plugin_wakatime_apiurl` to https://wakatime.com and let users override it with their self-hosted instance if needed 🙂 This way we don't even need a  `plugin_wakatime_wakapi` option

File: source/plugins/wakatime/metadata.yml [hunk 4]
```diff
@@ -53,3 +52,22 @@ inputs:
...
+    description: Use WakaTime or self hosted Wakapi instance
+    type: boolean
+    default: no
+
+  # If you use the public wakapi instance the url 
+  # would be https://wakapi.dev
+  plugin_wakatime_apiurl:
+    description: Address where to reach your Wakapi instance
+    type: string
+    default: ""
```
Comment (lowlighter) [inline source/plugins/wakatime/metadata.yml:68]:
```suggestion
    default: http://wakatime.com
```
I would also prefer `plugin_wakatime_url` to match with other `url` options present in other plugins

File: source/plugins/wakatime/metadata.yml [hunk 5]
```diff
@@ -53,3 +52,22 @@ inputs:
     type: number
     default: 5
     min: 0
+
+  # If you use a selfhosted wakatime instance
+  # that is publicly available set this to true
+  plugin_wakatime_wakapi:
+    description: Use WakaTime or self hosted Wakapi instance
+    type: boolean
+    default: no
```
Comment (lowlighter) [inline source/plugins/wakatime/metadata.yml:61]: Can be deleted as we can merge into a single option with `plugin_wakatime_apiurl`

**Review by lowlighter — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution 👍 !
I requested some changes but nothing major 🙂

**Comment by MeerBiene** [comment]
I think I got everything resolved, let me know if otherwise :)

File: source/plugins/wakatime/metadata.yml (hunk 1 above)
Comment (MeerBiene) [reply source/plugins/wakatime/metadata.yml]: ohh thats bigbrain

File: source/plugins/wakatime/index.mjs [hunk 6]
```diff
@@ -1,44 +1,47 @@
...
+    //Check if plugin is enabled and requirements are met
+    if (!enabled || !q.wakatime) return null;
 
-        //Load inputs
-          let {sections, days, limit} = imports.metadata.plugins.wakatime.inputs({data, account, q})
-          if (!limit)
-            limit = void(limit)
-          const range = {"7":"last_7_days", "30":"last_30_days", "180":"last_6_months", "365":"last_year"}[days] ?? "last_7_days"
+    //Load inputs
+    let { sections, days, limit, apiurl, user } = imports.metadata.plugins.wakatime.inputs({ data, account, q });
```
Comment (lowlighter) [inline source/plugins/wakatime/index.mjs:9]:
```suggestion
    let { sections, days, limit, url, user } = imports.metadata.plugins.wakatime.inputs({ data, account, q });
```

File: source/plugins/wakatime/index.mjs [hunk 7]
```diff
@@ -1,44 +1,47 @@
...
-              daily:stats.daily_average/(60*60),
-            },
-            projects:stats.projects.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            languages:stats.languages.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            os:stats.operating_systems.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-            editors:stats.editors.map(({name, percent, total_seconds:total}) => ({name, percent:percent/100, total})).sort((a, b) => b.percent - a.percent).slice(0, limit),
-          }
+    //Querying api and format result (https://wakatime.com/developers#stats)
+    console.debug(`metrics/compute/${login}/plugins > wakatime > querying api`);
+    const {data: { data: stats }} = await imports.axios.get(`${apiurl}/api/v1/users/${user}/stats/${range}?api_key=${token}`);
```
Comment (lowlighter) [inline source/plugins/wakatime/index.mjs:21]:
```suggestion
    const {data: { data: stats }} = await imports.axios.get(`${url}/api/v1/users/${user}/stats/${range}?api_key=${token}`);
```

**Review by lowlighter — CHANGES_REQUESTED** [review]
Almost! 
Since `plugin_wakatime_apiurl` has been renamed to `plugin_wakatime_url` the inputs parser will send back `url` instead of `apiurl` 🙂 

After this it should be ok 👍 
(builds are currently broke because I'm experimenting stuff with vercel so it's ok)

**Comment by lowlighter** [comment]
Thanks!

**Pull request merged**